module/module.go

```diff
@@ -0,0 +1,422 @@
+	parentAddr              string
+	activeBackgroundWorkers sync.WaitGroup
+	handlers                HandlerMap
+	services                map[resource.Subtype]subtype.Service
+	pb.UnimplementedModuleServiceServer
+}
+
+// NewModule returns the basic module framework/structure.
+func NewModule(address string, logger *zap.SugaredLogger) *Module {
+	m := &Module{
+		logger:     logger,
+		addr:       address,
+		grpcServer: grpc.NewServer(),
+		ready:      true,
+		handlers:   make(HandlerMap),
+		services:   make(map[resource.Subtype]subtype.Service),
+	}
+	m.buildHandlerMap()
+	pb.RegisterModuleServiceServer(m.grpcServer, m)
+	reflection.Register(m.grpcServer)
+	return m
+}
+
+// Start starts the module service and grpc server.
+func (m *Module) Start(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	err := m.initSubtypeServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	oldMask := syscall.Umask(0o077)
+	lis, err := net.Listen("unix", m.addr)
+	syscall.Umask(oldMask)
+	if err != nil {
+		return errors.WithMessage(err, "failed to listen")
+	}
+
+	m.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer m.activeBackgroundWorkers.Done()
+		defer utils.UncheckedErrorFunc(func() error { return os.Remove(m.addr) })
+		m.logger.Infof("server listening at %v", lis.Addr())
+		if err := m.grpcServer.Serve(lis); err != nil {
+			m.logger.Fatalf("failed to serve: %v", err)
+		}
+	})
+	return nil
+}
+
+// Close shuts down the module and grpc server.
+func (m *Module) Close() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.logger.Info("Sutting down gracefully.")
+	m.grpcServer.GracefulStop()
+	m.activeBackgroundWorkers.Wait()
+}
+
+// GRPCServer returns the underlying grpc.Server instance for the module.
+func (m *Module) GRPCServer() *grpc.Server {
+	return m.grpcServer
+}
+
+// GetParentComponent returns a component from the parent robot by name.
+func (m *Module) GetParentComponent(ctx context.Context, name resource.Name) (interface{}, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.parent == nil {
+		rc, err := client.New(ctx, "unix://"+m.parentAddr, m.logger)
+		if err != nil {
+			return nil, err
+		}
+		m.parent = rc
+	}
+	return m.parent.ResourceByName(name)
+}
+
+// SetReady can be set to false if the module is not ready (ex. waiting on hardware).
+func (m *Module) SetReady(ready bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.ready = ready
+}
+
+// Ready receives the parent address and reports api/model combos the module is ready to service.
+func (m *Module) Ready(ctx context.Context, req *pb.ReadyRequest) (*pb.ReadyResponse, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.parentAddr = req.GetParentAddress()
+
+	return &pb.ReadyResponse{
+		Ready:      m.ready,
+		Handlermap: m.handlers.ToProto(),
+	}, nil
+}
+
+// AddComponent receives the component configuration from the parent.
+func (m *Module) AddComponent(ctx context.Context, req *pb.AddComponentRequest) (*pb.AddComponentResponse, error) {
```

Comment:
This function also relies heavily on the registry concept and our internal subtypeService system. May need to be rethought for other SDKs. E.g. a simple module might simply have the module author set a callback and then handle this directly.